scripts/visualize/visualize_utils.R

```diff
@@ -0,0 +1,160 @@
+  proj.string = CRS(proj4string(sp))
+  return(as.sp_box(xs, ys, proj.string))
+}
+
+
+#' create a spatial polygon from x and y coordinates
+#' 
+#' @param xs a numeric vector of length two, containing the min and max of x values
+#' @param ys a numeric vector of length two, containing the min and max of y values
+#' 
+#' @return a `SpatialPolygons` object
+as.sp_box <- function(xs, ys, proj.string){
+  Sr1 <- Polygon(cbind(c(xs[c(1, 2, 2, 1, 1)]), c(ys[c(1, 1, 2, 2, 1)])))
+  Srs1 <- Polygons(list(Sr1), "s1")
+  SpP <- SpatialPolygons(list(Srs1), proj4string = proj.string)
+  return(SpP)
+}
+
+#' extract the svg elements from an sp object
+#' 
+#' @param sp a spatial object
+#' @param ... additional arguments passed to the plotting methods of `sp` (e.g., `expandBB`)
+#' @param width the width (in inches) of the svg view
+#' @param height the height (in inches) of the svg view
+#' @param pointsize number of pixels per inch
+#' @param xlim x limits (in sp units) of the plot
+#' @param ylim y limits (in sp units) of the plot
+#' 
+#' @return an `xml_document`
+get_svg_geoms <- function(sp, ..., width = 10, height = 8, pointsize = 12, xlim, ylim){
+  
+  stopifnot(packageVersion('svglite') == '1.2.0.9002')
+  
+  if (missing(xlim)){
+    xlim <- get_sp_lims(sp, ..., return = 'xlim')
+  }
+  if (missing(ylim)){
+    ylim <- get_sp_lims(sp, ..., return = 'ylim')
+  }
+  
+  # clip the spatial object so that it only contains features and data that are within the plotting range:
+  clipped.sp <- clip_sp(sp, xlim, ylim)
+  
+  rendered <- svglite::xmlSVG(width = width, height = height, pointsize = pointsize, standalone = F, {
+    set_sp_plot()
+    plot(clipped.sp, ..., xlim = xlim, ylim = ylim)
+  })
+  
+  sp.geoms <- tail(xml2::xml_children(rendered), length(sp))
+  
+  # here either strip the important attributes out and re-add them with a xml_set_attrs call, or lapply the nodeset and add attrs one by one:
+  
+  return(sp.geoms) # removing the <rect/> element...
+}
+
+#' extract the plotting limits from a spatial object, given a sized svg view
+#' 
+#' @param sp a spatial object
+#' @param ... additional arguments passed to the plotting methods of `sp` (e.g., `expandBB`)
+#' @param width the width (in inches) of the svg view
+#' @param height the height (in inches) of the svg view
+#' @param pointsize number of pixels per inch
+#' @param return what limits to return
+#' 
+#' @return a list or numeric vector, depending on what is specified for `return`
+
+get_sp_lims <- function(sp, ..., width = 10, height = 8, pointsize = 12, return = c('xlim','ylim')){
+  bb <- get_sp_bbox(sp)
+  # now default plot
+  # extract usr, return lims from usr
+  .fun <- svglite::svgstring(width = width, height = height, pointsize = pointsize, standalone = F)
+  suppressWarnings(plot(bb, ...)) # warning is for expandBB param, if used
+  usr <- par('usr')
+  dev.off()
+  xlim = usr[c(1,2)]
+  ylim = usr[c(3,4)]
+  
+  list.out <- list(xlim = xlim, ylim = ylim)
```

Comment:
it seems like there's a lot of switching between xlim/ylim and bbox formats in this file. would it be feasible to use just one (i'd suggest bbox) throughout, only reformatting them where necessary within functions? i.e., this function could return a bbox format, and get_svg_geoms could accept bbox format. doesn't make a big difference for this one file, but might add up over time to have only one R object to deal with rather than 2 to pass around...that said, you can probably predict the value better than I at this point

Reply:
agreed, but the trick is that bbox is the bounds of the `sp` object, while `xlim` and `ylim` are the bounds of the actual plotting view (not same as `bbox`, since the plot view usually has a different aspect than the bbox). I think there is a smarter way to do this (agreed), but wanted to point that out.

Reply:
i did see that they're different things - i'm only proposing that you use a similar format, not change the numbers. and i also think it might make sense to just sit on this idea rather than doing anything about it for now...

Reply:
ahh, ok.